core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala

```diff
@@ -0,0 +1,201 @@
+    val (frameWithPodSpecChanges, changedPodIds) = frame.applySpecEvent(msg)
+    frame = frameWithPodSpecChanges
+    val frameEffects = SchedulerLogic.computeNextStateForPods(frameWithPodSpecChanges)(changedPodIds)
+    applyAndUpdateCaches(frameEffects, changedPodIds)
+  }
+
+  /**
+    * Given a frame effects, apply the usi state change effects of the frame effects. Then, update the cache and perform
+    * other housekeeping items for pods that are dirty, such as:
+    *
+    * - Prune terminal / unreachable podStatuses for which no podSpec is defined
+    * - Update the pending launch set index
+    * - (WIP) issue any revive calls (this should be done elsewhere)
+    *
+    * Any usiState changes (such as pruning terminal / unreachable podStatuses) are added to the frameEffects.
+    *
+    * @param eventFrameEffects The effects from the incoming SpecEvent
+    * @param dirtyPodIds The podSpecs whose cache should be updated or checked for pruning
+    * @return The frame effects with additional housekeeping effects applied
+    */
+  private def applyAndUpdateCaches(eventFrameEffects: FrameEffects, dirtyPodIds: Set[PodId]): FrameEffects = {
+    frame = frame.applyStateEffects(eventFrameEffects)
+    this.cachedPendingLaunch = this.cachedPendingLaunch.update(frame, dirtyPodIds)
+
+    val pruneEffects = SchedulerLogic.pruneTaskStatuses(frame)(dirtyPodIds)
+    frame = frame.applyStateEffects(pruneEffects)
+
+    val reviveEffects =
+      if (cachedPendingLaunch.pendingLaunch.nonEmpty)
+        FrameEffects.empty.withMesosCall(Mesos.Call.Revive)
+      else
+        FrameEffects.empty
+
+    eventFrameEffects ++ pruneEffects ++ reviveEffects
+  }
+
+  /**
+    * Process a Mesos event and update internal state.
+    *
+    * @param event
+    * @return The events describing state changes as Mesos call intents
+    */
+  def processMesosEvent(event: Mesos.Event): FrameEffects = {
+    val result = SchedulerLogic.handleMesosEvent(frame, cachedPendingLaunch.pendingLaunch)(event)
+
+    applyAndUpdateCaches(result.frameEffects, result.dirtyPodIds)
+  }
+}
+
+/**
+  * The current home for USI business logic
+  */
+private[core] object SchedulerLogic {
+  private def terminalOrUnreachable(status: PodStatus): Boolean = {
+    // TODO - temporary stub implementation
+    status.taskStatuses.values.forall(status => status == Mesos.TaskStatus.TASK_RUNNING)
+  }
+
+  private def taskIdsFor(pod: PodSpec): Seq[TaskId] = {
+    // TODO - temporary stub implementation
+    Seq(pod.id.value)
+  }
+
+  /**
+    * Process the modification of some podSpec.
+    *
+    * - If a podSpec is deleted:
+    *   - Also delete the podRecord.
+    *   - Prune a terminal / unknown task status.
+    * - If a podSpec is marked as terminal, then issue a kill.
+    *
+    * @return The effects of launching this offer
+    */
+  private def matchOffer(offer: Mesos.Event.Offer, pendingLaunchPodSpecs: Seq[PodSpec]): FrameEffects = {
+    var effects = FrameEffects.empty
```

Comment:
Wdyt about moving this already into a dedicated `OfferMatcher` class with some interface and one (naive) first-come-first-served implementation. This would show where we want to go with it eventually.

Reply:
yeah, not opposed to doing that later.